internal/controller/postgrescluster/volumes.go

```diff
@@ -122,6 +131,555 @@ func (r *Reconciler) observePersistentVolumeClaims(
+// with existing pgData volumes to ensure consistent naming and proper
+// bootstrapping.
+func (r *Reconciler) configureExistingPVCs(
+	ctx context.Context, cluster *v1beta1.PostgresCluster,
+	volumes []corev1.PersistentVolumeClaim,
+) ([]corev1.PersistentVolumeClaim, error) {
+
+	var err error
+
+	if cluster.Spec.DataSource != nil &&
+		cluster.Spec.DataSource.ExistingVolumes != nil &&
+		cluster.Spec.DataSource.ExistingVolumes.ExistingPGDataVolume != nil {
+		// If the startup instance name isn't set, use the instance set defined at position zero.
+		if cluster.Status.StartupInstance == "" {
+			set := &cluster.Spec.InstanceSets[0]
+			cluster.Status.StartupInstanceSet = set.Name
+			cluster.Status.StartupInstance = naming.GenerateStartupInstance(cluster, set).Name
+		}
+		volumes, err = r.configureExistingPGVolumes(ctx, cluster, volumes,
+			cluster.Status.StartupInstance)
+
+		// existing WAL volume must be paired with an existing pgData volume
+		if cluster.Spec.DataSource != nil &&
+			cluster.Spec.DataSource.ExistingVolumes != nil &&
+			cluster.Spec.DataSource.ExistingVolumes.ExistingPGWALVolume != nil &&
+			err == nil {
+			volumes, err = r.configureExistingPGWALVolume(ctx, cluster, volumes,
+				cluster.Status.StartupInstance)
+		}
+	}
+
+	if cluster.Spec.DataSource != nil &&
+		cluster.Spec.DataSource.ExistingVolumes != nil &&
+		cluster.Spec.DataSource.ExistingVolumes.ExistingPGBackRestVolume != nil &&
+		err == nil {
+
+		volumes, err = r.configureExistingRepoVolumes(ctx, cluster, volumes)
+	}
+	return volumes, err
+}
+
+// +kubebuilder:rbac:groups="",resources=persistentvolumeclaims,verbs=create;patch
+
+// configureExistingPGVolumes first searches the observed volumes list to see
+// if the existing pgData volume defined in the spec is already updated. If not,
+// this sets the appropriate labels and ownership for the volume to be used in
+// the PostgresCluster.
+func (r *Reconciler) configureExistingPGVolumes(
+	ctx context.Context,
+	cluster *v1beta1.PostgresCluster,
+	volumes []corev1.PersistentVolumeClaim,
+	instanceName string,
+) ([]corev1.PersistentVolumeClaim, error) {
+
+	// if the volume is already in the list, move on
+	for i := range volumes {
+		if cluster.Spec.DataSource.ExistingVolumes.ExistingPGDataVolume.
+			PVCName == volumes[i].Name {
+			return volumes, nil
+		}
+	}
+
+	if len(cluster.Spec.InstanceSets) > 0 {
+		if volName := cluster.Spec.DataSource.ExistingVolumes.
+			ExistingPGDataVolume.PVCName; volName != "" {
+			volume := &corev1.PersistentVolumeClaim{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      volName,
+					Namespace: cluster.Namespace,
+				},
+				Spec: cluster.Spec.InstanceSets[0].DataVolumeClaimSpec,
+			}
+
+			volume.ObjectMeta.Labels = map[string]string{
+				naming.LabelCluster:     cluster.Name,
+				naming.LabelInstanceSet: cluster.Spec.InstanceSets[0].Name,
+				naming.LabelInstance:    instanceName,
+				naming.LabelRole:        naming.RolePostgresData,
+			}
```

Comment:
That's a good question, I'm not sure. My aim was to add just the labels needed for the migration. 

Since the volumes aren't being created, they would retain whatever labels/annotations they had previously, but adding other custom labels may be desirable here. Curious if anyone else has thoughts.